HB/common/phant-abbreviation.elpi

```diff
@@ -1,73 +1,143 @@
+%   [Notation N x_0 .. x_n := C x_0 .. _ _ id .. x_i .. _ id _ _ id]
 % with a variable number of [_] between each [id], and where
 % - [x_i] is given by the user
 % - [_]   correspond to arguments that are left implicit,
 % - [id]  trigger unification as described in
-% /Canonical Structures for the working Coq user/ by Mahboubi and Tassi
+% - [Phant x] to infer the canonical structure on x
 %
-% phant-arg encode these three kind of arguments
-% - [x_i] is encoded using [real-arg x_i]
-% - [_]              using [implicit-arg]
-% - [id]             using [unify-arg]
-kind phant-arg type.
-type real-arg name -> phant-arg.
-type infer-type name -> phant-arg.
-type implicit-arg phant-arg.
-type unify-arg phant-arg.
+% See /Canonical Structures for the working Coq user/ by Mahboubi and Tassi
+
+% This type is private, build it via the APIs below
+typeabbrev phant-term phant.private.phant-term.
+
+namespace phant {
+
+% [add-abbreviation Name PhT C A] builds a definition "phant_Name" for the
+% term T and an abbreviation Name := phant_Name phantom-arguments as per Ph.
+% Use the API below to build a PhT as you like.
+pred add-abbreviation i:string, i:phant-term, o:constant, o:abbreviation.
+add-abbreviation N (private.phant-term AL T1) C Abbrev :- std.do! [
+  NC is "phant_" ^ N,
+  std.assert-ok! (coq.elaborate-skeleton T1 TTy T) "add-abbreviation: T illtyped",
+  log.coq.env.add-const-noimplicits NC T TTy @transparent! C,
+  private.add-abbreviation.term 0 (global (const C)) AL NParams AbbrevT,
+  @global! => log.coq.notation.add-abbreviation N NParams AbbrevT tt Abbrev,
+].
+
+% [of-gref GR ExplicitMixins PT] builds a phant-term taking all parameters,
+% the type, then inferring automatically all classing covering the mixins
+% GR depends on. ExplicitMixins is a list of mixins one wants to explicitly
+% pass (instead of being inferred)
+pred of-gref i:gref, i:list mixinname, o:phant-term.
+of-gref GRF RealMixinArgs PhBody:- !, std.do! [
+  std.assert! (gref-deps GRF MLwP) "mk-phant-term: unknown gref",
+  std.assert! (coq.env.typeof GRF FTy) "mk-phant-term: F illtyped",
+  coq.mk-eta (-1) FTy (global GRF) EtaF,
+  % toposort-mixins ML MLSorted,
+  MLwP = MLwPSorted, % Assumes we give them already sorted in dep order.
+  std.rev {list-w-params_list MLwPSorted} MLSortedRev,
+
+  std.map RealMixinArgs (m\r\ r = private.this-mixin-is-real-arg m) RMClauses,
+  std.filter MLSortedRev (m\not(std.mem! RealMixinArgs m)) MLSortedRevFiltered,
+  find-max-classes MLSortedRevFiltered CNL,
+  assert-good-coverage! MLSortedRevFiltered CNL,
+
+  RMClauses => w-params.then MLwP real-arg real-arg
+    (private.mk-phant-term.classes EtaF CNL) PhBody,
+].
+
+% A la carte API: start with a term and wrap it up
+
+% A term with no phantom arguments
+pred init i:term, o:phant-term.
+init T (private.phant-term [] T).
+
+% [real-arg N T Ph Ph1] Adds a real argument named N of type T around Ph
+pred real-arg i:name, i:term, i:(term -> phant-term), o:phant-term.
```

Comment:
I liked that the name was `.*fun.*`, because it has the signature of `fun` (except it's predicate).